flow/project.py

```diff
@@ -479,24 +468,502 @@ def complete(self, job):
+            pass
+
+        # the directives used will be {'nranks': 4} for op1
+        # and {'nranks': 2, 'executable': 'python3'} for op2
+    """
+
+    MAX_LEN_ID = 100
+
+    def __init__(self, name, operations=None, operation_directives=None,
+                 options=""):
+        self.name = name
+        self.options = options
+        self.operations = dict() if operations is None else operations
+        if operation_directives is None:
+            self.operation_directives = dict()
+        else:
+            self.operation_directives = operation_directives
+
+    def _set_entrypoint_item(self, entrypoint, directives, key, default, job):
+        """Set a value (executable, path) for entrypoint in command.
+
+        Order of priority is the operation directives specified and
+        then the project specified value.
+        """
+        entrypoint[key] = directives.get(key, entrypoint.get(key, default))
+        if callable(entrypoint[key]):
+            entrypoint[key] = entrypoint[key](job)
+
+    def _determine_entrypoint(self, entrypoint, directives, job):
+        """Get the entrypoint for creating a JobOperation.
+
+        If path cannot be determined, then raise a RuntimeError since we do not
+        know where to point to.
+        """
+        entrypoint = entrypoint.copy()
+        self._set_entrypoint_item(entrypoint, directives, 'executable', sys.executable, job)
+        self._set_entrypoint_item(entrypoint, directives, 'path', None, job)
+        if entrypoint['path'] is None:
+            raise RuntimeError("Entrypoint path not set.")
+        else:
+            return "{} {}".format(entrypoint['executable'], entrypoint['path']).lstrip()
+
+    def _resolve_directives(self, name, defaults):
+        if name in self.operation_directives:
+            return deepcopy(self.operation_directives[name])
+        else:
+            return deepcopy(defaults.get(name, dict()))
+
+    def _submit_cmd(self, entrypoint, ignore_conditions, parallel, job=None):
+        entrypoint = self._determine_entrypoint(entrypoint, dict(), job)
+        cmd = "{} run -o {}".format(entrypoint, self.name)
+        cmd = cmd if job is None else cmd + ' -j {}'.format(job)
+        cmd = cmd if not parallel else cmd + " --parallel"
+        cmd = cmd if self.options is None else cmd + ' ' + self.options
+        if ignore_conditions != IgnoreConditions.NONE:
+            return cmd.strip() + ' --ignore-conditions=' + str(ignore_conditions)
+        else:
+            return cmd.strip()
+
+    def _run_cmd(self, entrypoint, operation_name, operation, directives, job):
+        if isinstance(operation, FlowCmdOperation):
+            return operation(job).lstrip()
+        else:
+            entrypoint = self._determine_entrypoint(entrypoint, directives, job)
+            return '{} exec {} {}'.format(entrypoint, operation_name, job).lstrip()
+
+    def __iter__(self):
+        yield from self.operations.values()
+
+    def __repr__(self):
```

Comment:
This function fails, because the constructor creates different attributes now.

- [ ] Add unit test for `FlowGroup.__repr__`
- [ ] Fix `Flowgroup.__repr__`.